### DISCUSSION TITLE
RealTime Flutter Chat Sample (or any sample)

### DISCUSSION THREAD

**Discussion by jtkeyva** [supabase/supabase#6037] [Questions] [answered]
I have been unable to find any examples or resources getting realtime working with Flutter. I ran the example from the realtime_flutter / realtime-dart package but there was no documentation so I couldn't get it to work.

I feel a Chat demo would go a long way if anyone has any resources.

In the meantime can someone please help with a simple chunk of code that listens to INSERT DELETE on a page? Could be as simple as a list. I can manually add/delete stuff from Supabase UI to test, so don't need all bells and whistles.

Thanks!

**Comment by w3b6x9** [comment]
@jtkeyva I think @dshukertjr might be able to provide some info / help.

**Comment by dshukertjr** [comment, marked as answer]
@jtkeyva I am so sorry to take a real long time to create this example, but I have created a early version of a chat sample app here:
https://github.com/supabase-community/flutter-chat

I haven't added much descriptive comments to it, but I plan to add some more docs and will probably create a video around it as well!

**Reply by jtkeyva** [reply]
I really appreciate this as a starting point to a full fledged chat app! Yes a video would be an awesome addition as well as Supabase Admin Panel settings and setting up callbacks & URL schemes.

Group chat would be a great next step as well as user presence/online/offline/status. I look forward to seeing what you come up with and will happily test and provide feedback.

I noticed changing a username was realtime as well which is a nice touch. Is it practical to have real-time username updates in a full scale production app?

**Reply by dshukertjr** [reply]
@jtkeyva That is a very good question. I think it could very much depend on the detailed nature of the chat app, but depending on the scale of the app, it might be a better idea to not listen to information that does not update very often such as user names.

**Reply by jtkeyva** [reply]
@dshukertjr got it thanks!

**Reply by 9ermemo** [reply]
sir can you explain us more about realtime and stream with streamBuilder in flutter

**Reply by dshukertjr** [reply]
@9ermemo Sure, what would you like to know about realtime and stream with StreamBuilder in Flutter? Their difference?

**Reply by 9ermemo** [reply]
yes Sir i need just a few exemples like how to use realtime with stream builder in flutter , especially this function 
supabase.channel('*').on(
  RealtimeListenTypes.postgresChanges,
  ChannelFilter(event: '*', schema: '*'),
  (payload, [ref]) {
    print('Change received: ${payload.toString()}');
  },
).subscribe();
 
you knew that stream builder take  Stream but this method return void , and i don't knew how to handle this , 

we need some vidoes explain that cuz the community of supabase is  small  i can't find more about it on internet 
thank you sir have a good day

**Reply by dshukertjr** [reply]
@9ermemo 
I am so glad you mentioned video examples! I plan to work on some Flutter + Supabase video content starting soon, so will for sure cover these realtime stuff there!

In the mean while, let me clarify the difference between `.stream()` and `.channel().subscribe()`.

`.stream()` and `.channel().subscribe()` are both methods in supabase-flutter to listen to realtime events from your Supabase instance. The main difference is that `.stream()` gives you the initial data where as `.channel().subscribe()` will only give you the change since you started listening. In most cases `.stream()` should be good enough to be used to listen to realtime data. 

If you ever want more finer control over the realtime data, you can use `.channel().subscribe()`. Since `.channe.().subscribe()` has a callback, you can call `setState` for example inside the callback to get the data and display it in your UI.

```dart
supabase.channel('channel_name').on(
  RealtimeListenTypes.postgresChanges,
  ChannelFilter(event: '', schema: '*'),
  (payload, [ref]) {
    setState(() {
      data = payload['new'];
    });
  },
).subscribe();
```

**Reply by 9ermemo** [reply]
Oh thank thank you sir so much , it's my fault  , i have use .channel().subscribe(). and i didnt see the change in UI cuz i forgot to use setstate method for seeing the change ,  thank sir very much for your help and your support
